internal/net/tcp/dialer_test.go

```diff
@@ -503,90 +970,458 @@ func Test_dialer_DialContext(t *testing.T) {
+					}
+					return nil
+				},
+			}
+		}(),
+		func() test {
+			addr := "google.com"
+
+			cache, _ := cache.New()
+			cache.Set(addr, &dialerCache{
+				ips: []string{
+					"invalid_ip",
+				},
+			})
+
+			return test{
+				name: "retry when cache invalid and cache will be deleted",
+				args: args{
+					dctx:    context.Background(),
+					network: "tcp",
+					addr:    addr + ":80",
+				},
+				fields: fields{
+					der: &net.Dialer{
+						Resolver: &net.Resolver{
+							PreferGo: false,
+						},
+					},
+					cache:    cache,
+					dnsCache: true,
+				},
+				checkFunc: func(d *dialer, w want, gotConn net.Conn, err error) error {
+					if err != nil {
+						return errors.Errorf("err is not nil: %v", err)
+					}
+					if gotConn == nil {
+						return errors.New("conn is nil")
+					}
+
+					if c, ok := cache.Get(addr); ok {
+						return errors.Errorf("cache value is set: %+v", c)
+					}
+					return nil
+				},
+			}
+		}(),
+		func() test {
+			srvNums := 30
+			srvs := make([]*httptest.Server, 0, srvNums)
+			hosts := make([]string, 0, srvNums)
+			ports := make([]string, 0, srvNums)
+
+			for i := 0; i < srvNums; i++ {
+				srvs = append(srvs, httptest.NewServer(http.HandlerFunc(
+					func(w http.ResponseWriter, r *http.Request) {
+						fmt.Fprint(w, fmt.Sprint(i))
+						w.WriteHeader(200)
+					},
+				)))
+				hp := srvs[i].URL[7:len(srvs[i].URL)]
+				h, p, _ := net.SplitHostPort(hp)
+				hosts = append(hosts, h)
+				ports = append(ports, fmt.Sprint(p))
+			}
+
+			addr := "address"
+
+			cache, _ := cache.New()
+			cache.Set(addr, &dialerCache{
+				ips: hosts,
+			})
+
+			return test{
+				name: "return cached ip connection in round robin order",
+				args: args{
+					dctx:    context.Background(),
+					network: "tcp",
+					addr:    addr + ":" + ports[0],
+				},
+				fields: fields{
+					der: &net.Dialer{
+						Resolver: &net.Resolver{
+							PreferGo: false,
+						},
+					},
+					cache:    cache,
+					dnsCache: true,
+				},
+				checkFunc: func(d *dialer, w want, gotConn net.Conn, err error) error {
+					c, _ := d.cache.Get(addr)
+					dc := c.(*dialerCache)
+
+					check := func(gotConn net.Conn, gotErr error, cnt int, port string) error {
+						defer func() {
+							_ = gotConn.Close()
+						}()
+
+						if gotErr != nil {
+							return errors.Errorf("err is not nil: %v", gotErr)
+						}
+						if gotConn == nil {
+							return errors.New("conn is nil")
+						}
+						if c := atomic.LoadUint32(&dc.cnt); c != uint32(cnt) {
+							return errors.Errorf("cnt not correct, %d, except: %d", c, cnt)
+						}
+
+						// check the connection made is the same excepted
+						_, p, _ := net.SplitHostPort(gotConn.RemoteAddr().String())
+						if fmt.Sprint(p) != port {
+							return errors.Errorf("unexcepted port number, except: %v, got: %v", port, p)
+						}
+
+						// read the output from the server and check if it is equals to the count
+						/*
```

Comment:
I think this comment out lines are not needed

Reply:
I wanted to test the connection returned from the `cachedDial()` are working or not but I couldn't find a way to test it. (@kpango suggested using pipe but seems it is not working)

as Vald is working for now I wanna to skip this test for now and work in the future, so I left this comment here.

Reply:
well, so could you add `TODO` comment the above?
I'm afraid we'll forget implementing about it.

Reply:
```suggestion
						/* TODO verify the connection is working
```

Reply:
how about using starting test TCP listener?